File: android_app/app/src/main/java/com/oceansentinels/app/presentation/viewmodel/MeshViewModel.kt
package com.oceansentinels.app.presentation.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.oceansentinels.app.domain.model.HazardType
import com.oceansentinels.app.domain.model.UrgencyLevel
import com.oceansentinels.app.mesh.ble.BleMeshManager
import com.oceansentinels.app.mesh.model.*
import com.oceansentinels.app.mesh.repository.MeshMessageRepository
import com.oceansentinels.app.mesh.service.MeshForegroundService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * ViewModel for the Mesh Network screen.
 * Manages mesh service state, message queue, and peer list.
 */
@HiltViewModel
class MeshViewModel @Inject constructor(
    application: Application,
    private val meshRepository: MeshMessageRepository,
    private val bleMeshManager: BleMeshManager
) : AndroidViewModel(application) {

    companion object {
        private const val TAG = "MeshViewModel"
    }

    // ==================== UI State ====================

    private val _meshStatus = MutableStateFlow(MeshNetworkStatus())
    val meshStatus: StateFlow<MeshNetworkStatus> = _meshStatus.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _sendState = MutableStateFlow<MeshSendState>(MeshSendState.Idle)
    val sendState: StateFlow<MeshSendState> = _sendState.asStateFlow()

    private val _selectedTab = MutableStateFlow(MeshTab.QUEUE)
    val selectedTab: StateFlow<MeshTab> = _selectedTab.asStateFlow()

    // ==================== Data Flows ====================

    /** All mesh messages */
    val allMessages: StateFlow<List<MeshMessage>> = meshRepository.getAllMessages()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    /** Own pending messages */
    val pendingMessages: StateFlow<List<MeshMessage>> = meshRepository.getPendingMessages()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    /** Delivered messages */
    val deliveredMessages: StateFlow<List<MeshMessage>> = meshRepository.getDeliveredMessages()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    /** Messages relayed for others */
    val relayedMessages: StateFlow<List<MeshMessage>> = meshRepository.getRelayedMessages()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    /** Service running state */
    val isServiceRunning: StateFlow<Boolean> = MeshForegroundService.isServiceRunning

    init {
        updateMeshCapabilities()
        observeCounters()
    }

    // ==================== Mesh Service Control ====================

    /** Start the mesh foreground service */
    fun startMeshService() {
        val context = getApplication<Application>()
        MeshForegroundService.start(context)
        updateStatus { it.copy(isRunning = true) }
    }

    /** Stop the mesh foreground service */
    fun stopMeshService() {
        val context = getApplication<Application>()
        MeshForegroundService.stop(context)
        updateStatus { it.copy(isRunning = false) }
    }

    /** Toggle mesh service on/off */
    fun toggleMeshService() {
        if (isServiceRunning.value) {
            stopMeshService()
        } else {
            startMeshService()
        }
    }

    // ==================== Report Hazard via Mesh ====================

    /**
     * Submit a hazard report through the mesh network.
     * Auto-determines transport: Internet → Mesh → Local Queue
     */
    fun reportHazard(
        hazardType: HazardType,
        location: String,
        latitude: Double?,
        longitude: Double?,
        description: String,
        urgency: UrgencyLevel,
        contactInfo: String? = null,
        photoUrl: String? = null,
        reporterUserId: Int? = null
    ) {
        viewModelScope.launch {
            _sendState.value = MeshSendState.Sending

            val result = meshRepository.createAndSend(
                hazardType = hazardType,
                location = location,
                latitude = latitude,
                longitude = longitude,
                description = description,
                urgency = urgency,
                contactInfo = contactInfo,
                photoUrl = photoUrl,
                reporterUserId = reporterUserId
            )

            result.fold(
                onSuccess = { message ->
                    // Use the actual transport stored on the message rather than
                    // inferring from status (Issue #12: prevents incorrect mapping)
                    _sendState.value = MeshSendState.Success(
                        messageId = message.messageId,
                        transport = when (message.status) {
                            MeshMessageStatus.DELIVERED -> MeshTransport.INTERNET
                            MeshMessageStatus.RELAYED -> MeshTransport.BLE_CODED
                            else -> MeshTransport.LOCAL_QUEUE
                        },
                        status = message.status
                    )
                    Timber.i("$TAG: Report submitted: ${message.messageId} via ${message.status}")
                },
                onFailure = { error ->
                    _sendState.value = MeshSendState.Error(error.message ?: "Failed to submit report")
                    Timber.e(error, "$TAG: Failed to submit report")
                }
            )
        }
    }

    /** Reset the send state back to idle */
    fun resetSendState() {
        _sendState.value = MeshSendState.Idle
    }

    // ==================== Tab Selection ====================

    fun selectTab(tab: MeshTab) {
        _selectedTab.value = tab
    }

    // ==================== Capabilities ====================

    private fun updateMeshCapabilities() {
        updateStatus {
            it.copy(
                isBleAvailable = bleMeshManager.isBleAvailable(),
                isCodedPhySupported = bleMeshManager.isCodedPhySupported()
            )
        }

        // Observe reactive peer counts and BLE state from the singleton BleMeshManager
        viewModelScope.launch {
            bleMeshManager.connectedPeerCount.collect { count ->
                updateStatus { it.copy(connectedPeerCount = count) }
            }
        }
        viewModelScope.launch {
            bleMeshManager.discoveredPeerCount.collect { count ->
                updateStatus { it.copy(discoveredPeerCount = count) }
            }
        }
        viewModelScope.launch {
            bleMeshManager.isRunningFlow.collect { running ->
                updateStatus { it.copy(isRunning = running) }
            }
        }
        viewModelScope.launch {
            bleMeshManager.isAdvertisingFlow.collect { adv ->
                updateStatus { it.copy(isAdvertising = adv) }
            }
        }
        viewModelScope.launch {
            bleMeshManager.isScanningFlow.collect { scan ->
                updateStatus { it.copy(isScanning = scan) }
            }
        }
    }

    private fun observeCounters() {
        viewModelScope.launch {
            allMessages.collect { messages ->
                updateStatus {
                    it.copy(
                        pendingMessageCount = messages.count {
                            m -> m.status in listOf(MeshMessageStatus.PENDING, MeshMessageStatus.SENDING, MeshMessageStatus.FAILED)
                        },
                        deliveredMessageCount = messages.count { m -> m.status == MeshMessageStatus.DELIVERED },
                        relayedMessageCount = messages.count { m -> m.hopCount > 0 }
                    )
                }
            }
        }
    }

    private fun updateStatus(update: (MeshNetworkStatus) -> MeshNetworkStatus) {
        _meshStatus.value = update(_meshStatus.value)
    }

    fun clearError() {
        _error.value = null
    }
}

/**
 * State for mesh message sending
 */
sealed class MeshSendState {
    data object Idle : MeshSendState()
    data object Sending : MeshSendState()
    data class Success(
        val messageId: String,
        val transport: MeshTransport,
        val status: MeshMessageStatus
    ) : MeshSendState()
    data class Error(val message: String) : MeshSendState()
}

/**
 * Tabs for the mesh screen
 */
enum class MeshTab(val title: String) {
    ALL("All"),
    QUEUE("Queue"),
    DELIVERED("Delivered"),
    RELAYED("Relayed"),
    PEERS("Peers")
}

File: IOS/OceanSentinels/Presentation/ViewModels/MeshViewModel.swift
import Foundation
import Observation
import Combine

// MARK: - MeshViewModel

/// ViewModel for the Mesh Network screen.
/// Manages mesh service state, message queue, and peer list.
@Observable
@MainActor
final class MeshViewModel {

    // MARK: - UI State

    var meshStatus: MeshNetworkStatus = MeshNetworkStatus()
    var isLoading: Bool = false
    var error: String?
    var sendState: MeshSendState = .idle
    var selectedTab: MeshTab = .queue

    // MARK: - Data

    var allMessages: [MeshMessage] = []
    var pendingMessages: [MeshMessage] = []
    var deliveredMessages: [MeshMessage] = []
    var relayedMessages: [MeshMessage] = []

    // MARK: - Dependencies

    private let meshRepository: MeshMessageRepository
    private let bleMeshManager: BleMeshManager

    // Observation tasks
    private var observationTasks: [Task<Void, Never>] = []

    // MARK: - Init

    init(meshRepository: MeshMessageRepository, bleMeshManager: BleMeshManager) {
        self.meshRepository = meshRepository
        self.bleMeshManager = bleMeshManager

        updateMeshCapabilities()
        startObservation()
    }

    deinit {
        observationTasks.forEach { $0.cancel() }
    }

    // MARK: - Mesh Service Control

    /// Start the mesh background service.
    func startMeshService() {
        MeshBackgroundService.shared.startMesh()
        meshStatus.isRunning = true
    }

    /// Stop the mesh background service.
    func stopMeshService() {
        MeshBackgroundService.shared.stopMesh()
        meshStatus.isRunning = false
    }

    /// Toggle mesh service on/off.
    func toggleMeshService() {
        if meshStatus.isRunning {
            stopMeshService()
        } else {
            startMeshService()
        }
    }

    // MARK: - Report Hazard via Mesh

    /// Submit a hazard report through the mesh network.
    /// Auto-determines transport: Internet → Mesh → Local Queue
    func reportHazard(
        hazardType: HazardType,
        location: String,
        latitude: Double?,
        longitude: Double?,
        description: String,
        urgency: UrgencyLevel,
        contactInfo: String? = nil,
        photoUrl: String? = nil,
        reporterUserId: Int? = nil
    ) {
        Task {
            sendState = .sending

            let result = await meshRepository.createAndSend(
                hazardType: hazardType,
                location: location,
                latitude: latitude,
                longitude: longitude,
                description: description,
                urgency: urgency,
                contactInfo: contactInfo,
                photoUrl: photoUrl,
                reporterUserId: reporterUserId
            )

            switch result {
            case .success(let message):
                let transport: MeshTransport = switch message.status {
                case .delivered: .internet
                case .relayed: .bleCoded
                default: .localQueue
                }

                sendState = .success(
                    messageId: message.messageId,
                    transport: transport,
                    status: message.status
                )
                AppLogger.mesh.info("Report submitted: \(message.messageId) via \(message.status.rawValue)")

            case .failure(let error):
                sendState = .error(error.localizedDescription)
                AppLogger.mesh.error("Failed to submit report: \(error.localizedDescription)")
            }
        }
    }

    /// Reset the send state back to idle.
    func resetSendState() {
        sendState = .idle
    }

    // MARK: - Tab Selection

    func selectTab(_ tab: MeshTab) {
        selectedTab = tab
    }

    // MARK: - Capabilities & Observation

    private func updateMeshCapabilities() {
        meshStatus.isBleAvailable = bleMeshManager.isBleAvailable()
        meshStatus.isCodedPhySupported = bleMeshManager.isCodedPhySupported()
    }

    private func startObservation() {
        // Observe connected peer count via Combine publisher
        let peerTask = Task { [weak self] in
            guard let self else { return }
            for await count in bleMeshManager.$connectedPeerCount.values {
                self.meshStatus.connectedPeerCount = count
            }
        }
        observationTasks.append(peerTask)

        // Observe discovered peer count
        let discoveredTask = Task { [weak self] in
            guard let self else { return }
            for await count in bleMeshManager.$discoveredPeerCount.values {
                self.meshStatus.discoveredPeerCount = count
            }
        }
        observationTasks.append(discoveredTask)

        // Observe running state
        let runningTask = Task { [weak self] in
            guard let self else { return }
            for await running in bleMeshManager.$isRunning.values {
                self.meshStatus.isRunning = running
            }
        }
        observationTasks.append(runningTask)

        // Observe advertising state
        let advertisingTask = Task { [weak self] in
            guard let self else { return }
            for await advertising in bleMeshManager.$isAdvertising.values {
                self.meshStatus.isAdvertising = advertising
            }
        }
        observationTasks.append(advertisingTask)

        // Observe scanning state
        let scanningTask = Task { [weak self] in
            guard let self else { return }
            for await scanning in bleMeshManager.$isScanning.values {
                self.meshStatus.isScanning = scanning
            }
        }
        observationTasks.append(scanningTask)

        // Periodically refresh message lists
        let messageTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                await self.refreshMessages()
                try? await Task.sleep(for: .seconds(5))
            }
        }
        observationTasks.append(messageTask)
    }

    private func refreshMessages() async {
        let all = await meshRepository.getAllMessages()
        let pending = all.filter { $0.status == .pending || $0.status == .sending || $0.status == .failed }
        let delivered = all.filter { $0.status == .delivered }
        let relayed = all.filter { $0.hopCount > 0 }

        await MainActor.run {
            self.allMessages = all
            self.pendingMessages = pending
            self.deliveredMessages = delivered
            self.relayedMessages = relayed

            self.meshStatus.pendingMessageCount = pending.count
            self.meshStatus.deliveredMessageCount = delivered.count
            self.meshStatus.relayedMessageCount = relayed.count
        }
    }

    func clearError() {
        error = nil
    }
}

// MARK: - MeshSendState

/// State for mesh message sending.
enum MeshSendState: Equatable {
    case idle
    case sending
    case success(messageId: String, transport: MeshTransport, status: MeshMessageStatus)
    case error(String)
}

// MARK: - MeshTab

/// Tabs for the mesh screen.
enum MeshTab: String, CaseIterable {
    case all = "All"
    case queue = "Queue"
    case delivered = "Delivered"
    case relayed = "Relayed"
    case peers = "Peers"

    var title: String { rawValue }
}
